### PR TITLE
0.25: add deploy receipt / add block data to receipt / change estimate fee input

### PR DESCRIPTION
**Pull request by ArielElp**

### CODE REVIEW

**Review by lior-stark — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/22)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @anatgstarkware and @ArielElp)

___
*[`api/starknet_api_openrpc.json` line 429 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GCDA-4_3ZK3NpxjFk:-N7GCDA-4_3ZK3NpxjFl:b-1lp03t) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L429)):*
> ```JSON
>                     "summary": "The transaction to estimate",
>                     "schema": {
>                         "$ref": "#/components/schemas/INVOKE_TXN"
> ```

why is this?
this means a call to estimate fee will need to provide things like max_fee, nonce and signature.
___
*[`api/starknet_api_openrpc.json` line 1258 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GAqFUDN4EpxbwyByr:-N7GAqFUDN4EpxbwyBys:b-3ztvyr) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L1258)):*
> ```JSON
>                         "description": "Extra information pertaining to the status"
>                     },
>                     "block_hash": {
> ```

why not use the block_id object, instead of 2 separate properties for number and hash?
___
*[`api/starknet_api_openrpc.json` line 1315 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GAxzS9cOYM54REdSJ:-N7GAxzS9cOYM54REdSK:bgjvb5x) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L1315)):*
> ```JSON
>             "DEPLOY_TXN_RECEIPT": {
>                 "title": "Deploy Transaction Receipt",
>                 "allOf": [
> ```

why do you need `allOf` if you have a single schema reference?

so basically `DEPLOY_TXN_RECEIPT` is an alias of `COMMON_RECEIPT_PROPERTIES`?

**Review by ArielElp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/22)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @anatgstarkware and @lior-stark)

___
*[`api/starknet_api_openrpc.json` line 429 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GCDA-4_3ZK3NpxjFk:-N7GD1NYDN47rrZQZv0k:b-tuyqox) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L429)):*
<details><summary><i>Previously, lior-stark wrote…</i></summary><blockquote>

why is this?
this means a call to estimate fee will need to provide things like max_fee, nonce and signature.
</blockquote></details>

Yep, this is all needed to estimate fee currently (validate has access to max_fee, so you have to specify it). Not ideal, but this is a broader StarkNet discussion. 

So far we decided that we're ok with it, and we will rename this endpoint to something like "simulate" in 0.10, which will return all the data we have (possible trace) + fee estimation. I didn't want to burden the spec with this at this point, but we do need to change the schema as right now the endpoint is not usable (I can't estimate fee on txs from my account, I have to break up the call).
___
*[`api/starknet_api_openrpc.json` line 1258 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GAqFUDN4EpxbwyByr:-N7GCmxB5w2YdOEjd6LX:bcek1ia) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L1258)):*
<details><summary><i>Previously, lior-stark wrote…</i></summary><blockquote>

why not use the block_id object, instead of 2 separate properties for number and hash?
</blockquote></details>

I want them both in the response (why not + also the case in Eth)
block_id is one_of, and block_tag makes no sense in the response
___
*[`api/starknet_api_openrpc.json` line 1315 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GAxzS9cOYM54REdSJ:-N7GDm9A9bOk36ADRhLM:bavt89j) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L1315)):*
<details><summary><i>Previously, lior-stark wrote…</i></summary><blockquote>

why do you need `allOf` if you have a single schema reference?

so basically `DEPLOY_TXN_RECEIPT` is an alias of `COMMON_RECEIPT_PROPERTIES`?
</blockquote></details>

Yes, same as declare_receipt

removed allOf (not sure if this is valid now)

**Review by lior-stark — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/22)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @anatgstarkware and @ArielElp)

___
*[`api/starknet_api_openrpc.json` line 429 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GCDA-4_3ZK3NpxjFk:-N7GG0u2AVyh9hhAQGDA:b-tdu3vl) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L429)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Yep, this is all needed to estimate fee currently (validate has access to max_fee, so you have to specify it). Not ideal, but this is a broader StarkNet discussion. 

So far we decided that we're ok with it, and we will rename this endpoint to something like "simulate" in 0.10, which will return all the data we have (possible trace) + fee estimation. I didn't want to burden the spec with this at this point, but we do need to change the schema as right now the endpoint is not usable (I can't estimate fee on txs from my account, I have to break up the call).

</blockquote></details>

what's the `simulate` change exactly?
let's discuss tomorrow.
___
*[`api/starknet_api_openrpc.json` line 1315 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GAxzS9cOYM54REdSJ:-N7GFoAu5cNJLGjlv4zw:bckglvo) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L1315)):*
<details><summary><i>Previously, ArielElp wrote…</i></summary><blockquote>

Yes, same as declare_receipt

removed allOf (not sure if this is valid now)
</blockquote></details>

than i think we should fix the DECLARE_RECEIPT as well.

I'm still seeing the `allOf`, did you push the change?

**Review by ArielElp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/22)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @anatgstarkware and @lior-stark)

___
*[`api/starknet_api_openrpc.json` line 429 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GCDA-4_3ZK3NpxjFk:-N7GIQp-6GMvUIvGKgSv:b-w9c36v) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L429)):*
<details><summary><i>Previously, lior-stark wrote…</i></summary><blockquote>

what's the `simulate` change exactly?
let's discuss tomorrow.
</blockquote></details>

makes the response richer
___
*[`api/starknet_api_openrpc.json` line 1315 at r2](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-N7GAxzS9cOYM54REdSJ:-N7GILLL7oOFtpl1paT1:b4ceo45) ([raw file](https://github.com/starkware-libs/starknet-specs/blob/ce9fb416e14195bd3c538dca1e4c361b8e15c8ce/api/starknet_api_openrpc.json#L1315)):*
<details><summary><i>Previously, lior-stark wrote…</i></summary><blockquote>

than i think we should fix the DECLARE_RECEIPT as well.

I'm still seeing the `allOf`, did you push the change?
</blockquote></details>

It sees it now, fixed declare as well

**Review by lior-stark — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/22)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @anatgstarkware)

**Review by lior-stark — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/starknet-specs/22#-:-N7LRdjV4tD6vh6LHroM:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @anatgstarkware)
